### PR TITLE
Stable release 2.0.7

### PR DESCRIPTION
**Pull request by Dead2**
Changes since 2.0.6:
- Fix CVE-2022-37434 #1328
- Fix chunkmemset #1196
- Fix deflateBound too small #1236
- Fix Z_SOLO #1263
- Fix ACLE variant of crc32 #1274
- Fix inflateBack #1311
- Fix deflate_quick windowsize #1431 
- Fix DFLTCC bugs related to adler32 #1349 and #1390
- Fix warnings #1194 #1312 #1362
- MacOS build fix #1198
- Add invalid windowBits handling #1293
- Support for Force TZCNT #1186
- Support for aligned_alloc() #1360
- Minideflate improvements #1175 #1238
- Dont use unaligned access for memcpy #1309
- Build system #1209 #1233 #1267 #1273 #1278 #1292 #1316 #1318 #1365
- Test improvements #1208 #1227 #1241 #1353
- Cleanup #1266
- Documentation #1205 #1359
- Misc improvements #1294 #1297 #1306 #1344 #1348
- Backported zlib fixes
- Backported CI workflows from Develop branch

### CODE REVIEW

**Comment by phprus** [comment]
Missing commit:
https://github.com/zlib-ng/zlib-ng/pull/1267 - fix for bug in https://github.com/zlib-ng/zlib-ng/pull/1393/commits/3a5a2c436c7dc3d05b041b8f1dd99302b910ff8f

**Comment by Dead2** [comment]
> Missing commit: #1267 - fix for bug in [3a5a2c4](https://github.com/zlib-ng/zlib-ng/commit/3a5a2c436c7dc3d05b041b8f1dd99302b910ff8f)

What do you mean? The commit is in this PR already, and there is no other commit in the source PR.

Edit: Actually the commit is already there. But github does not show it in the above list, only in the commits tab. The list above seems to only show about half the actual commits, weird..

**Comment by phprus** [comment]
Oh! I am sorry!
For some unknown reason, not all commits and changes (on "Files changed" tab) were loaded for me.

**Comment by mtl1979** [comment]
Build breaks initially at [9bb926a](https://github.com/zlib-ng/zlib-ng/pull/1393/commits/9bb926a89178e83dce733e0312a21d1daced4c99) ... After that everything is noise.

That commit use zmemcpy_* macros that were not available.

Next break seems to be at [0ce59bf](https://github.com/zlib-ng/zlib-ng/pull/1393/commits/0ce59bfb0a231c3dbe3f5a6591a52b060b2f5933)... That's where tests start to fail if I drop the commit above.

File: deflate.h [hunk 1]
```diff
@@ -290,79 +290,58 @@ typedef enum {
...
-    *(uint16_t *)(&s->pending_buf[s->pending]) = w;
-    s->pending += 2;
-#else
-    put_byte(s, (w & 0xff));
-    put_byte(s, ((w >> 8) & 0xff));
+#if BYTE_ORDER == BIG_ENDIAN
+    w = ZSWAP16(w);
 #endif
+    memcpy(&s->pending_buf[s->pending], &w, sizeof(w));
 }
```
Comment (mtl1979) [inline deflate.h]: Something is missing here?

```
    memcpy(&s->pending_buf[s->pending], &w, sizeof(w));
    s->pending += 2;
}
```

**Review by mtl1979 — CHANGES_REQUESTED** [review]
Broken `put_short()`.

File: test/minideflate.c [hunk 2]
```diff
@@ -279,21 +291,60 @@ int main(int argc, char **argv) {
...
+        if (uncompr)
+            window_bits += 32;
+    }
+
+    if (window_bits == INT32_MAX) {
+        window_bits = MAX_WBITS;
+        /* Auto-detect wrapper for inflateInit */
+        if (uncompr)
+            window_bits += 32;
+    }
```
Comment (mtl1979) [inline test/minideflate.c]: This block seems duplicated. See commits [c587cb9](https://github.com/zlib-ng/zlib-ng/pull/1393/commits/c587cb9e11be18ece1bf35e42cfc9cbb03b7bdb1) and [dc05f44](https://github.com/zlib-ng/zlib-ng/pull/1393/commits/dc05f44398f081761aa8c93dd069f4348a449f14).

**Comment by Dead2** [comment]
@mtl1979 Thanks for the assist 😄.
The offending commits have now been edited and repushed.

**Comment by mtl1979** [comment]
> @mtl1979 Thanks for the assist 😄. The offending commits have now been edited and repushed.

I've said many times that we need a new stable release soon... I will do anything within the limits of my knowledge...

There still seems to be issues finding clang-11 and Visual C++ compiler...

**Comment by Dead2** [comment]
> There still seems to be issues finding clang-11 and Visual C++ compiler...

Those commits would have to be re-made from scratch, as the CMake code has diverged quite a lot since then.

PS: Also updated the PR description.

**Comment by mtl1979** [comment]
> > There still seems to be issues finding clang-11 and Visual C++ compiler...
> 
> Those commits would have to be re-made from scratch, as the CMake code has diverged quite a lot since then.
> 
> PS: Also updated the PR description.

I think we should force the failing runs to use "ubuntu-20.04" and "windows-2019" so the fix stays simple... Trying to update the build system to work with latest versions might take too long.

**Comment by iii-i** [comment]
Could you please pull https://github.com/iii-i/zlib-ng/tree/stable-pre-release?

https://github.com/iii-i/zlib-ng/commit/f0a02c0f9055350bb6cc498ba32fb3ac73340aac is a combination of https://github.com/zlib-ng/zlib-ng/pull/1349 and https://github.com/zlib-ng/zlib-ng/pull/1390 (not merged yet) minus tests.

**Comment by rhpvorderman** [comment]
I tested the prerelease against the test suite of python-zlib-ng. The following issue remains: https://github.com/zlib-ng/zlib-ng/issues/1417

It is fixed on the develop branch.

**Comment by Dead2** [comment]
Getting close to having the CI passing now.
I  am working on getting the abicheck files updated, but the sanitizers are complaining about a memory leak.

Ex:
https://github.com/zlib-ng/zlib-ng/actions/runs/4126285777/jobs/7128014363
```
682/692 Test #668: CVE-2004-0797 ..............................................................................................................***Failed    0.69 sec
/home/runner/work/zlib-ng/zlib-ng/minigzip: <fd:0>: invalid bit length repeat

=================================================================
==13496==ERROR: LeakSanitizer: detected memory leaks

Indirect leak of 16384 byte(s) in 1 object(s) allocated from:
    #0 0x7fa5b37da55c in __interceptor_posix_memalign ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:226
    #1 0x558e2ac36b5d in gz_look (/home/runner/work/zlib-ng/zlib-ng/minigzip+0x20b5d)

Indirect leak of 8192 byte(s) in 1 object(s) allocated from:
    #0 0x7fa5b37da55c in __interceptor_posix_memalign ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:226
    #1 0x558e2ac36a8c in gz_look (/home/runner/work/zlib-ng/zlib-ng/minigzip+0x20a8c)

Indirect leak of 7160 byte(s) in 1 object(s) allocated from:
    #0 0x7fa5b37da55c in __interceptor_posix_memalign ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:226
    #1 0x558e2ac61135 in zng_calloc (/home/runner/work/zlib-ng/zlib-ng/minigzip+0x4b135)
    #2 0x558e2ac7ab21  (/home/runner/work/zlib-ng/zlib-ng/minigzip+0x64b21)

Indirect leak of 232 byte(s) in 1 object(s) allocated from:
    #0 0x7fa5b37da55c in __interceptor_posix_memalign ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:226
    #1 0x558e2ac3417e in gz_open (/home/runner/work/zlib-ng/zlib-ng/minigzip+0x1e17e)

Indirect leak of 7 byte(s) in 1 object(s) allocated from:
    #0 0x7fa5b37d9867 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x558e2ac34723 in gz_open (/home/runner/work/zlib-ng/zlib-ng/minigzip+0x1e723)

SUMMARY: AddressSanitizer: 31975 byte(s) leaked in 5 allocation(s).
CMake Error at cmake/run-and-redirect.cmake:53 (message):
  /home/runner/work/zlib-ng/zlib-ng/minigzip;-d failed: Subprocess aborted
```

Weirdly, only CVE tests fail..
```
The following tests FAILED:
	667 - CVE-2002-0059 (Failed)
	668 - CVE-2004-0797 (Failed)
	669 - CVE-2005-1849 (Failed)
	670 - CVE-2005-2096 (Failed)
```

Any ideas?
I tried to have a look at the commit log for gzlib.c and gzguts.h, but the two commits missing should not matter.
Coverity flags no problems.

**Comment by nmoinvaz** [comment]
Perhaps #1422 should go in since some other related changes went in too..

**Comment by Dead2** [comment]
> Perhaps #1422 should go in since some other related changes went in too..

This branch predates even the zmemcmp macros, so all the changes would have to be remade from scratch.

**Comment by nmoinvaz** [comment]
Hmm I think you are right, I guess I was mistaken.

**Comment by Dead2** [comment]
Regarding the memory leaks, I am at a loss.
I am unable to reproduce any of them locally.

`gz_look()` allocates `state->in` and `state->out`, and `gzclose_r()` is supposed to deallocate them.
Those two variables fit well well as they are 8KiB and 16KiB, matching the first two leaks.
But `gzlib.c` and `gzread.c` are practically the same as they are in `develop`, where it does not leak.
`Minigzip` calls those functions from its `gz_uncompress()` function, nothing different from `develop` there either.

The rest of the leaks I have not looked at.

I am coming down with the flu or something, been getting steadily worse since this morning, so I might not be around much or at all the next few days. Trying to button up what I can while I am still able to look at the monitor.

**Comment by Dead2** [comment]
@mtl1979 Package check fails one of the runs, it looks like something went completely wrong and it is missing output from one of the runs, or something like that?
https://github.com/zlib-ng/zlib-ng/actions/runs/4129678522/jobs/7135548390

**Comment by Dead2** [comment]
@iii-i One of the DFLTCC runs is failing:
Seems to succeed on GCC but fail with Clang.
https://github.com/zlib-ng/zlib-ng/actions/runs/4129678532/jobs/7135553202
```
==4648==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x2aa2b95bd9d in dfltcc_can_deflate (/home/actions-runner/_work/zlib-ng/zlib-ng/switchlevels+0xdbd9d)
    #1 0x2aa2b941867 in zng_deflateBound (/home/actions-runner/_work/zlib-ng/zlib-ng/switchlevels+0xc1867)
    #2 0x2aa2b934811 in main (/home/actions-runner/_work/zlib-ng/zlib-ng/switchlevels+0xb4811)
    #3 0x3ff90f24409 in __libc_start_main (/lib/s390x-linux-gnu/libc.so.6+0x24409)
    #4 0x2aa2b8a11d3 in _start (/home/actions-runner/_work/zlib-ng/zlib-ng/switchlevels+0x211d3)
```

**Comment by mtl1979** [comment]
> @mtl1979 Package check fails one of the runs, it looks like something went completely wrong and it is missing output from one of the runs, or something like that? https://github.com/zlib-ng/zlib-ng/actions/runs/4129678522/jobs/7135548390

I think one of the builds is targeting wrong processor...

File: .github/workflows/pkgcheck.yml [hunk 3]
```diff
@@ -15,63 +15,71 @@ jobs:
...
 
-          - name: Ubuntu GCC ARM HF
+          - name: Ubuntu GCC ARM SF
             os: ubuntu-latest
-            chost: arm-linux-gnueabihf
-            compiler: arm-linux-gnueabihf-gcc
-            cmake-args: -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain-arm.cmake -DCMAKE_C_COMPILER_TARGET=arm-linux-gnueabihf
-            packages: qemu gcc-arm-linux-gnueabihf libc6-dev-armhf-cross
+            cmake-args: -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain-arm.cmake
+            packages: qemu gcc-arm-linux-gnueabi g++-arm-linux-gnueabi libc-dev-armel-cross
```
Comment (mtl1979) [inline .github/workflows/pkgcheck.yml]: `chost:` and `compiler:`were removed here instead of correcting them.

Comment (Dead2) [reply .github/workflows/pkgcheck.yml]: Right, it was changed to be similar to the ARM SF targets from cmake.yml, will fix.

**Comment by Dead2** [comment]
Only things blocking the release now are the ASAN and MSAN CI failures.

**Comment by nmoinvaz** [comment]
I did a BeyondCompare between the two branches and didn't notice very much difference in many source files including gz source files. It would be good to get in all the Mark Adler fixes: e713f2f4, 4af454bf, 33e671b9, 0e49aa38, 5393c219, 6581fb0e, 24d7d0d7.

**Comment by mtl1979** [comment]
ASAN leak seems to be coming from line `error(PREFIX(gzerror)(in, &err));` in `gz_uncompress()` of `test/minigzip.c`. `in` is not closed properly.

**Comment by iii-i** [comment]
For MSan, please take https://github.com/iii-i/zlib-ng/commit/2050ed59b8697d1b9258c473ed12f2398409648b and https://github.com/iii-i/zlib-ng/commit/4b32f7c5c489e2f49a08c1a03272f18aca3e5182 from https://github.com/iii-i/zlib-ng/commits/stable-msan.

**Comment by iii-i** [comment]
Sorry, I just noticed that I introduced a duplicate CI job (https://github.com/iii-i/zlib-ng/commit/4b32f7c5c489e2f49a08c1a03272f18aca3e5182#diff-cdd48abbd3eb8d1c54077449fc74a8de1f29805d2be5d8e5232b7aab76ea7a6fR238), and had to respin the branch. Please take https://github.com/iii-i/zlib-ng/commit/2050ed59b8697d1b9258c473ed12f2398409648b and https://github.com/iii-i/zlib-ng/commit/bba6bf1f8acdc009df7743492035a97e4496a42c instead.

**Comment by nmoinvaz** [comment]
All checks pass!

**Comment by mtl1979** [comment]
> All checks pass!

We're getting closer and closer to next stable release... Still need some testing before we can do official release...

**Comment by rhpvorderman** [comment]
I don't know if it is very helpful but for the python bindings I test the compatibility with proper zlib. I simply take all possible compression levels, memory levels and windowbits settings and run them through zlib-ng's deflate and see if zlib's inflate can properly handle them. Then I do the same thing the other way round. This results in thousands of compatibility tests.
I just checked this branch and on Windows x86/64 (msvc), Linux x86/64(gcc) and MacOS x86/64(clang) everything seems to be fine.

File: test/pkgcheck.sh [hunk 4]
```diff
@@ -125,7 +125,7 @@ cd btmp1
     ;;
   esac
   ../configure $CONFIGURE_ARGS
-  make
+  make -j2
```
Comment (rhpvorderman) [inline test/pkgcheck.sh:128]: Technically `make -j` (without any number) would also work here and leverage computers with higher core counts as well.

Comment (turol) [reply test/pkgcheck.sh:128]: This is a bad idea. It will start all the compilation jobs at once and can easily run out of memory.

Comment (rhpvorderman) [reply test/pkgcheck.sh:128]: Fair enough. It also does not really matter for the speed as the configuration steps are single-threaded and are the major bottlenecks.

**Comment by Dead2** [comment]
Heads up: This is getting released later today, unless something comes up in last minute testing.

I'd like it if people would do a quick review of this PR. (You don't need to read every line, just say what you did do; cursory review, tested on x86-64, agree with the proposed changes, etc).
(Merge protection is on, so reviews are also required before merging, and I'd like to avoid having to bypass those protections)

I meant to release this late last week, but I have been out with the flu for a week now, and I am finally able to spend a little time in front of a monitor again.

I have been running this on a production web server for a while now without problems, unfortunately I no longer have access to big server deployments since I am no longer employed (I became 100% disabled recently).
But confidence is high that there are no glaring bugs in this, but corner cases are always the hardest ones to test.

**Comment by Dead2** [comment]
GCC 12.2 on x86-64 i9-9900K (no AVX-512)

## 2.0.6
```
 Tool: minigzip   Levels: 1-9
 Runs: 45         Trim worst: 5

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     47.618%      1.852/1.881/1.900/0.011        0.897/0.921/0.936/0.010      100,930,174
 2     35.518%      2.716/2.745/2.766/0.013        0.990/1.006/1.021/0.007       75,282,961
 3     34.198%      3.106/3.157/3.206/0.025        0.953/0.979/0.990/0.008       72,485,726
 4     32.928%      3.562/3.622/3.670/0.031        0.922/0.935/0.947/0.007       69,793,995
 5     32.662%      3.817/3.869/3.938/0.034        0.908/0.927/0.938/0.007       69,228,858
 6     32.507%      4.308/4.412/4.473/0.047        0.908/0.927/0.938/0.007       68,901,209
 7     32.255%      5.925/6.066/6.185/0.072        0.887/0.908/0.923/0.009       68,366,800
 8     32.167%      8.655/8.787/9.027/0.092        0.894/0.908/0.920/0.007       68,180,776
 9     32.156%   11.603/11.802/12.225/0.151        0.876/0.907/0.921/0.010       68,156,160

 avg1  34.668%                        5.149                          0.935
 tot                               1853.663                        336.716      661,326,659

   text    data     bss     dec     hex filename
 112085    1296      32  113413   1bb05 libz-ng.so.2
```

## 2.0.7
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.407%      1.777/1.791/1.811/0.009        0.975/0.994/1.008/0.010       94,124,801
 2     35.518%      2.753/2.777/2.798/0.012        1.013/1.036/1.047/0.007       75,282,961
 3     34.198%      3.118/3.180/3.227/0.027        0.989/1.008/1.020/0.008       72,485,726
 4     32.928%      3.585/3.659/3.715/0.034        0.947/0.963/0.975/0.007       69,793,995
 5     32.662%      3.813/3.891/3.959/0.037        0.945/0.958/0.970/0.007       69,228,858
 6     32.507%      4.321/4.452/4.566/0.059        0.928/0.954/0.964/0.007       68,901,209
 7     32.255%      5.977/6.092/6.207/0.062        0.901/0.935/0.951/0.010       68,366,800
 8     32.167%      8.646/8.793/9.029/0.096        0.910/0.928/0.945/0.010       68,180,776
 9     32.156%   11.625/11.800/12.054/0.106        0.903/0.927/0.939/0.008       68,156,160

 avg1  34.311%                        5.159                          0.967
 tot                               1857.406                        348.108      654,521,286

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.407%      1.779/1.799/1.815/0.010        0.976/0.993/1.008/0.009       94,124,801
 2     35.518%      2.731/2.780/2.799/0.012        1.023/1.038/1.052/0.008       75,282,961
 3     34.198%      3.116/3.179/3.225/0.025        0.982/1.005/1.017/0.009       72,485,726
 4     32.928%      3.599/3.656/3.705/0.028        0.943/0.963/0.976/0.008       69,793,995
 5     32.662%      3.816/3.897/3.961/0.039        0.940/0.956/0.968/0.008       69,228,858
 6     32.507%      4.336/4.436/4.512/0.054        0.928/0.955/0.966/0.008       68,901,209
 7     32.255%      5.949/6.085/6.202/0.067        0.903/0.932/0.945/0.010       68,366,800
 8     32.167%      8.658/8.817/9.079/0.124        0.909/0.928/0.940/0.009       68,180,776
 9     32.156%   11.649/11.815/12.112/0.121        0.908/0.928/0.942/0.009       68,156,160

 avg1  34.311%                        5.163                          0.966
 tot                               1858.591                        347.925      654,521,286

   text    data     bss     dec     hex filename
 111613    1296      32  112941   1b92d libz-ng.so.2
 ```

Code size is slightly smaller. 
Compression and Decompression is slightly slower overall, but not by much.
Level 1 compression has improved by quite a bit, but it is also 4.5% faster.

**Comment by iii-i** [comment]
Thank you very much for the release! FWIW, I ran some additional tests on s390x, and all is looking good.

**Comment by rhpvorderman** [comment]
Thanks very much for releasing. I updated the [python bindings](https://github.com/pycompression/python-zlib-ng.git) to include the new version. Since the wbits bug is fixed these bindings can be used as a one-to-one replacement for python's zlib module.
